### PR TITLE
Tvb cosim

### PR DESCRIPTION
**Pull request by lionelkusch**
Add two monitors for interface with TVB : 

The simplest is Interface_co_simulation_parallel.py. The input is the state variable and the output is the state of other node and the input of proxy node. 

The second monitor is Interface_co_simulation.py. The input is intermediate result. This monitor is need some modification of the model in order to include this intermediate result.

### CODE REVIEW

**Comment by maedoc** [comment]
should I be able to test this with the current version of NEST or I need the latest/master?

**Comment by lionelkusch** [comment]
This version is independent of Nest or other simulator.
The tests are based on the TVB coupled with TVB or some specific function.

There are a proof of concept based on this interface and Nest. The proof of concept use MPI for the communication. In consequence, you need a wrapper over this interface for the MPI communication. 
This implementation is base on the repertory : https://gitlab.version.fz-juelich.de/diaz1/co-simulation-tvb-nest.git
The version of Nest is the following version : https://github.com/nest/nest-simulator/pull/1456
This pull request is the creation of Input and Output for Nest with MPI.
If you need an access to the proof of concept, you should send me an email ( lionel.kusch@univ-amu.fr)

**Comment by maedoc** [comment]
> This implementation is base on the repertory : https://gitlab.version.fz-juelich.de/diaz1/co-simulation-tvb-nest.git

@lionelkusch can you add me to that repo? I don't have access, user is `woodman1`

**Comment by lionelkusch** [comment]
I give you the access.

**Comment by lionelkusch** [comment]
I update master and fix a bug in one of interface

**Comment by lionelkusch** [comment]
The public project of the co-simulation between Nest and TVB is as the following link : 
https://github.com/multiscale-cosim/TVB-NEST

**Comment by liadomide** [comment]
This PR is not easy to inspect (too many files appear changed). It would have been better to rebased towards master.
Could you cherry pick your commits and put them on top of the current master ?

File: scientific_library/tvb/tests/library/simulator/test_monitor_cosimulation/co_simulation_paralelle/double_proxy_precision_complexe.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        print("Fail compare")
+        print(diff_2)
+        print(result_ref)
+        print(result_2)
+    else:
+        print('test' + str(i + 1) + ' : succeed')
+
+print("time_synchronize : %r s"%(time_synchronize*10**-3))
+print("max error S rate for 1 : %r" % max_error_s_1)
+print("max error S rate for 2: %r" % max_error_s_2)
```
[the remark this answers is not in the document]
Comment (lionelkusch) [reply scientific_library/tvb/tests/library/simulator/test_monitor_cosimulation/co_simulation_paralelle/double_proxy_precision_complexe.py:83]: Yes, it was some print for helping me to debug the code. I am modify all the test for be in the right format

File: scientific_library/tvb/tests/library/simulator/test_monitor_cosimulation/co_simulation_paralelle/double_proxy_precision_simple.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        print("Fail compare")
+        print(diff_2)
+        print(result_ref)
+        print(result_2)
+    else:
+        print('test' + str(i + 1) + ' : succeed')
+
+print("time_synchronize : %r s"%(time_synchronize*10**-3))
+print("max error S rate for 1 : %r" % max_error_s_1)
+print("max error S rate for 2: %r" % max_error_s_2)
```
[the remark this answers is not in the document]
Comment (lionelkusch) [reply scientific_library/tvb/tests/library/simulator/test_monitor_cosimulation/co_simulation_paralelle/double_proxy_precision_simple.py:109]: Most of these test can be also use for demonstration.
We need to decide what is the most appropriate for the demonstration.

File: scientific_library/tvb/tests/library/simulator/test_monitor_cosimulation/co_simulation_paralelle/function_tvb.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+        the time, the firing rate and the state of the network
+    """
+    if data_proxy is not None :
+        data_proxy[1] = np.reshape(data_proxy[1],(data_proxy[1].shape[0],data_proxy[1].shape[1],1,1))
+    result=sim.run(proxy_data=data_proxy,simulation_length=time)
+    time = result[0][0]
+    s = result[0][1][:,0]
+    return time,s
+
+class tvb_sim:
```
[the remark this answers is not in the document]
Comment (lionelkusch) [reply scientific_library/tvb/tests/library/simulator/test_monitor_cosimulation/co_simulation_paralelle/function_tvb.py:109]: Ok, I will refactore the code for the respect this convention

File: scientific_library/tvb/tests/library/simulator/test_monitor_cosimulation/co_simulation_paralelle/double_proxy_precision_simple.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (liadomide) [reply scientific_library/tvb/tests/library/simulator/test_monitor_cosimulation/co_simulation_paralelle/double_proxy_precision_simple.py:109]: I think both aspects should be covered.
under `tests` we should have automated checks, to be run with every change of the code, and ensure regressions don't happen (too often), while demo notebooks would also be great to have (under tvb_documentation would be my recommendation).

File: scientific_library/tvb/tests/library/simulator/test_monitor_cosimulation/co_simulation_paralelle/double_proxy_precision_complexe.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
+#  Copyright 2020 Forschungszentrum Jülich GmbH and Aix-Marseille Université
+# "Licensed to the Apache Software Foundation (ASF) under one or more contributor license agreements; and to You under the Apache License, Version 2.0. "
```
[the remark this answers is not in the document]
Comment (lionelkusch) [reply scientific_library/tvb/tests/library/simulator/test_monitor_cosimulation/co_simulation_paralelle/double_proxy_precision_complexe.py:2]: I forget to change the licence.

**Comment by lionelkusch** [comment]
I close the Pull Request because I don't know how to fix my error due to the update of master.
I create a new one (https://github.com/the-virtual-brain/tvb-root/pull/190) with all the comments already include for continuing.
Sorry for the inconvenient.

**Pull request closed without merging**